### PR TITLE
Clean up and Refactoring

### PR DESCRIPTION
**Pull request by blink1073**
- Use the 'nocursor' CodeMirror readonly state.
- Refactor so that each level of the hierarchy only interacts with its direct children.
- Clean up the CodeMirror modes.

### CODE REVIEW

File: src/notebook/codemirror-ipythongfm.ts [hunk 1]
```diff
@@ -15,6 +10,15 @@ import 'codemirror/mode/gfm/gfm';
...
+ * Define an IPython GFM (GitHub Flavored Markdown) mode. 
+ *
+ * Is just a slightly altered GFM Mode with support for latex.
+ * Latex support was supported by Codemirror GFM as of
+ *   https://github.com/codemirror/CodeMirror/pull/567
+ *  But was later removed in
+ *   https://github.com/codemirror/CodeMirror/commit/d9c9f1b1ffe984aee41307f3e927f80d1f23590c
+ */
 CodeMirror.defineMode("ipythongfm", (config: CodeMirror.EditorConfiguration, modeOptions?: any) => {
   let gfm_mode = CodeMirror.getMode(config, "gfm");
```
Comment (afshin) [inline src/notebook/codemirror-ipythongfm.ts]: `'gfm'`

File: src/notebook/codemirror-ipythongfm.ts [hunk 2]
```diff
@@ -15,6 +10,15 @@ import 'codemirror/mode/gfm/gfm';
...
+/**
+ * Define an IPython GFM (GitHub Flavored Markdown) mode. 
+ *
+ * Is just a slightly altered GFM Mode with support for latex.
+ * Latex support was supported by Codemirror GFM as of
+ *   https://github.com/codemirror/CodeMirror/pull/567
+ *  But was later removed in
+ *   https://github.com/codemirror/CodeMirror/commit/d9c9f1b1ffe984aee41307f3e927f80d1f23590c
+ */
 CodeMirror.defineMode("ipythongfm", (config: CodeMirror.EditorConfiguration, modeOptions?: any) => {
```
Comment (afshin) [inline src/notebook/codemirror-ipythongfm.ts]: `'ipythongfm'`

File: src/notebook/codemirror-ipythongfm.ts [hunk 3]
```diff
@@ -15,6 +10,15 @@ import 'codemirror/mode/gfm/gfm';
...
   let gfm_mode = CodeMirror.getMode(config, "gfm");
   let tex_mode = CodeMirror.getMode(config, "stex");
```
Comment (afshin) [inline src/notebook/codemirror-ipythongfm.ts]: `'stex'`

File: src/notebook/codemirror-ipythongfm.ts [hunk 4]
```diff
@@ -1,10 +1,5 @@
-// IPython GFM (GitHub Flavored Markdown) mode is just a slightly altered GFM
-// Mode with support for latex.
-//
-// Latex support was supported by Codemirror GFM as of
-//   https://github.com/codemirror/CodeMirror/pull/567
-// But was later removed in
-//   https://github.com/codemirror/CodeMirror/commit/d9c9f1b1ffe984aee41307f3e927f80d1f23590c
+// Copyright (c) Jupyter Development Team.
+// Distributed under the terms of the Modified BSD License.
 "use strict";
```
Comment (afshin) [inline src/notebook/codemirror-ipythongfm.ts]: `'use strict';`

**Comment by afshin** [comment]
There are a bunch of double quotes. I marked a few. Otherwise, :+1:

**Comment by dwillmer** [comment]
:+1:

**Pull request merged**